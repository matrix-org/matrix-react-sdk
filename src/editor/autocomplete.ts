/*
Copyright 2019 New Vector Ltd
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import {KeyboardEvent} from "react";

import {Part, CommandPartCreator, PartCreator} from "./parts";
import DocumentPosition from "./position";
import {ICompletion} from "../autocomplete/Autocompleter";
import Autocomplete from "../components/views/rooms/Autocomplete";

export interface ICallback {
    replaceParts?: Part[];
    close?: boolean;
}

export type UpdateCallback = (data: ICallback) => void;
export type GetAutocompleterComponent = () => Autocomplete;
export type UpdateQuery = (test: string) => Promise<void>;

export default class AutocompleteWrapperModel {
    private queryPart: Part;
    private partIndex: number;

    constructor(
        private updateCallback: UpdateCallback,
        private getAutocompleterComponent: GetAutocompleterComponent,
        private updateQuery: UpdateQuery,
        private partCreator: PartCreator | CommandPartCreator,
    ) {
    }

    public onEscape(e: KeyboardEvent) {
        this.getAutocompleterComponent().onEscape(e);
        this.updateCallback({
            replaceParts: [this.partCreator.plain(this.queryPart.text)],
            close: true,
        });
    }

    public close() {
        this.updateCallback({close: true});
    }

    public hasSelection() {
        return this.getAutocompleterComponent().hasSelection();
    }

    public hasCompletions() {
        const ac = this.getAutocompleterComponent();
        return ac && ac.countCompletions() > 0;
    }

    public onEnter() {
        this.updateCallback({close: true});
    }

    /**
     * If there is no current autocompletion, start one and move to the first selection.
     */
    public async startSelection() {
        const acComponent = this.getAutocompleterComponent();
        if (acComponent.countCompletions() === 0) {
            // Force completions to show for the text currently entered
            await acComponent.forceComplete();
            // Select the first item by moving "down"
            await acComponent.moveSelection(+1);
        }
    }

    public selectPreviousSelection() {
        this.getAutocompleterComponent().moveSelection(-1);
    }

    public selectNextSelection() {
        this.getAutocompleterComponent().moveSelection(+1);
    }

    public onPartUpdate(part: Part, pos: DocumentPosition) {
        // cache the typed value and caret here
        // so we can restore it in onComponentSelectionChange when the value is undefined (meaning it should be the typed text)
        this.queryPart = part;
        this.partIndex = pos.index;
        return this.updateQuery(part.text);
    }

    public onComponentSelectionChange(completion: ICompletion) {
        if (!completion) {
            this.updateCallback({
                replaceParts: [this.queryPart],
            });
        } else {
            this.updateCallback({
                replaceParts: this.partForCompletion(completion),
            });
        }
    }

    public onComponentConfirm(completion: ICompletion) {
        this.updateCallback({
            replaceParts: this.partForCompletion(completion),
            close: true,
        });
    }

    private partForCompletion(completion: ICompletion) {
        const {completionId} = completion;
        const text = completion.completion;
        switch (completion.type) {
            case "room":
                return [this.partCreator.roomPill(text, completionId), this.partCreator.plain(completion.suffix)];
            case "at-room":
                return [this.partCreator.atRoomPill(completionId), this.partCreator.plain(completion.suffix)];
            case "user":
                // Insert suffix only if the pill is the part with index 0 - we are at the start of the composer
                return this.partCreator.createMentionParts(this.partIndex === 0, text, completionId);
            case "command":
                // command needs special handling for auto complete, but also renders as plain texts
                return [(this.partCreator as CommandPartCreator).command(text)];
            default:
                // used for emoji and other plain text completion replacement
                return [this.partCreator.plain(text)];
        }
    }
}
