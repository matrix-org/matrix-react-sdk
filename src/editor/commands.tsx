/*
Copyright 2019 - 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { logger } from "matrix-js-sdk/src/logger";
import { IContent } from "matrix-js-sdk/src/models/event";

import EditorModel from "./model";
import { Type } from "./parts";
import { Command, CommandCategories, getCommand } from "../SlashCommands";
import { ITranslatableError, _t, _td } from "../languageHandler";
import Modal from "../Modal";
import ErrorDialog from "../components/views/dialogs/ErrorDialog";
import QuestionDialog from "../components/views/dialogs/QuestionDialog";

export function isSlashCommand(model: EditorModel): boolean {
    const parts = model.parts;
    const firstPart = parts[0];
    if (firstPart) {
        if (firstPart.type === Type.Command && firstPart.text.startsWith("/") && !firstPart.text.startsWith("//")) {
            return true;
        }

        if (firstPart.text.startsWith("/") && !firstPart.text.startsWith("//")
            && (firstPart.type === Type.Plain || firstPart.type === Type.PillCandidate)) {
            return true;
        }
    }
    return false;
}

export function getSlashCommand(model: EditorModel): [Command, string, string] {
    const commandText = model.parts.reduce((text, part) => {
        // use mxid to textify user pills in a command and room alias/id for room pills
        if (part.type === Type.UserPill || part.type === Type.RoomPill) {
            return text + part.resourceId;
        }
        return text + part.text;
    }, "");
    const { cmd, args } = getCommand(commandText);
    return [cmd, args, commandText];
}

export async function runSlashCommand(
    cmd: Command,
    args: string,
    roomId: string,
    threadId: string | null,
): Promise<[content: IContent | null, success: boolean]> {
    const result = cmd.run(roomId, threadId, args);
    let messageContent: IContent | null = null;
    let error = result.error;
    if (result.promise) {
        try {
            if (cmd.category === CommandCategories.messages || cmd.category === CommandCategories.effects) {
                messageContent = await result.promise;
            } else {
                await result.promise;
            }
        } catch (err) {
            error = err;
        }
    }
    if (error) {
        logger.error("Command failure: %s", error);
        // assume the error is a server error when the command is async
        const isServerError = !!result.promise;
        const title = isServerError ? _td("Server error") : _td("Command error");

        let errText;
        if (typeof error === 'string') {
            errText = error;
        } else if ((error as ITranslatableError).translatedMessage) {
            // Check for translatable errors (newTranslatableError)
            errText = (error as ITranslatableError).translatedMessage;
        } else if (error.message) {
            errText = error.message;
        } else {
            errText = _t("Server unavailable, overloaded, or something else went wrong.");
        }

        Modal.createDialog(ErrorDialog, {
            title: _t(title),
            description: errText,
        });
        return [null, false];
    } else {
        logger.log("Command success.");
        return [messageContent, true];
    }
}

export async function shouldSendAnyway(commandText: string): Promise<boolean> {
    // ask the user if their unknown command should be sent as a message
    const { finished } = Modal.createDialog(QuestionDialog, {
        title: _t("Unknown Command"),
        description: <div>
            <p>
                { _t("Unrecognised command: %(commandText)s", { commandText }) }
            </p>
            <p>
                { _t("You can use <code>/help</code> to list available commands. " +
                    "Did you mean to send this as a message?", {}, {
                    code: t => <code>{ t }</code>,
                }) }
            </p>
            <p>
                { _t("Hint: Begin your message with <code>//</code> to start it with a slash.", {}, {
                    code: t => <code>{ t }</code>,
                }) }
            </p>
        </div>,
        button: _t('Send as message'),
    });
    const [sendAnyway] = await finished;
    return sendAnyway;
}
