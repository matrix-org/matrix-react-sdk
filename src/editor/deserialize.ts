/*
Copyright 2019 New Vector Ltd
Copyright 2019, 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MatrixEvent } from "matrix-js-sdk/src/models/event";

import { walkDOMDepthFirst } from "./dom";
import { checkBlockNode } from "../HtmlUtils";
import { getPrimaryPermalinkEntity } from "../utils/permalinks/Permalinks";
import { PartCreator } from "./parts";

function parseAtRoomMentions(text: string, partCreator: PartCreator) {
    const ATROOM = "@room";
    const parts = [];
    text.split(ATROOM).forEach((textPart, i, arr) => {
        if (textPart.length) {
            parts.push(partCreator.plain(textPart));
        }
        // it's safe to never append @room after the last textPart
        // as split will report an empty string at the end if
        // `text` ended in @room.
        const isLast = i === arr.length - 1;
        if (!isLast) {
            parts.push(partCreator.atRoomPill(ATROOM));
        }
    });
    return parts;
}

function parseLink(a: HTMLAnchorElement, partCreator: PartCreator) {
    const {href} = a;
    const resourceId = getPrimaryPermalinkEntity(href); // The room/user ID
    const prefix = resourceId ? resourceId[0] : undefined; // First character of ID
    switch (prefix) {
        case "@":
            return partCreator.userPill(a.textContent, resourceId);
        case "#":
            return partCreator.roomPill(resourceId);
        default: {
            if (href === a.textContent) {
                return partCreator.plain(a.textContent);
            } else {
                return partCreator.plain(`[${a.textContent.replace(/[[\\\]]/g, c => "\\" + c)}](${href})`);
            }
        }
    }
}

function parseCodeBlock(n: HTMLElement, partCreator: PartCreator) {
    const parts = [];
    let language = "";
    if (n.firstChild && n.firstChild.nodeName === "CODE") {
        for (const className of (<HTMLElement>n.firstChild).classList) {
            if (className.startsWith("language-")) {
                language = className.substr("language-".length);
                break;
            }
        }
    }
    const preLines = ("```" + language + "\n" + n.textContent + "```").split("\n");
    preLines.forEach((l, i) => {
        parts.push(partCreator.plain(l));
        if (i < preLines.length - 1) {
            parts.push(partCreator.newline());
        }
    });
    return parts;
}

function parseHeader(el: HTMLElement, partCreator: PartCreator) {
    const depth = parseInt(el.nodeName.substr(1), 10);
    return partCreator.plain("#".repeat(depth) + " ");
}

interface IState {
    listIndex: number[];
    listDepth?: number;
}

function parseElement(n: HTMLElement, partCreator: PartCreator, lastNode: HTMLElement | undefined, state: IState) {
    switch (n.nodeName) {
        case "H1":
        case "H2":
        case "H3":
        case "H4":
        case "H5":
        case "H6":
            return parseHeader(n, partCreator);
        case "A":
            return parseLink(<HTMLAnchorElement>n, partCreator);
        case "BR":
            return partCreator.newline();
        case "EM":
            return partCreator.plain(`_${n.textContent}_`);
        case "STRONG":
            return partCreator.plain(`**${n.textContent}**`);
        case "PRE":
            return parseCodeBlock(n, partCreator);
        case "CODE":
            return partCreator.plain(`\`${n.textContent}\``);
        case "DEL":
            return partCreator.plain(`<del>${n.textContent}</del>`);
        case "LI": {
            const indent = "  ".repeat(state.listDepth - 1);
            if (n.parentElement.nodeName === "OL") {
                // The markdown parser doesn't do nested indexed lists at all, but this supports it anyway.
                const index = state.listIndex[state.listIndex.length - 1];
                state.listIndex[state.listIndex.length - 1] += 1;
                return partCreator.plain(`${indent}${index}. `);
            } else {
                return partCreator.plain(`${indent}- `);
            }
        }
        case "P": {
            if (lastNode) {
                return partCreator.newline();
            }
            break;
        }
        case "OL":
            state.listIndex.push((<HTMLOListElement>n).start || 1);
            /* falls through */
        case "UL":
            state.listDepth = (state.listDepth || 0) + 1;
            /* falls through */
        default:
            // don't textify block nodes we'll descend into
            if (!checkDescendInto(n)) {
                return partCreator.plain(n.textContent);
            }
    }
}

function checkDescendInto(node) {
    switch (node.nodeName) {
        case "PRE":
            // a code block is textified in parseCodeBlock
            // as we don't want to preserve markup in it,
            // so no need to descend into it
            return false;
        default:
            return checkBlockNode(node);
    }
}

function checkIgnored(n) {
    if (n.nodeType === Node.TEXT_NODE) {
        // riot adds \n text nodes in a lot of places,
        // which should be ignored
        return n.nodeValue === "\n";
    } else if (n.nodeType === Node.ELEMENT_NODE) {
        return n.nodeName === "MX-REPLY";
    }
    return true;
}

const QUOTE_LINE_PREFIX = "> ";
function prefixQuoteLines(isFirstNode, parts, partCreator) {
    // a newline (to append a > to) wouldn't be added to parts for the first line
    // if there was no content before the BLOCKQUOTE, so handle that
    if (isFirstNode) {
        parts.splice(0, 0, partCreator.plain(QUOTE_LINE_PREFIX));
    }
    for (let i = 0; i < parts.length; i += 1) {
        if (parts[i].type === "newline") {
            parts.splice(i + 1, 0, partCreator.plain(QUOTE_LINE_PREFIX));
            i += 1;
        }
    }
}

function parseHtmlMessage(html: string, partCreator: PartCreator, isQuotedMessage: boolean) {
    // no nodes from parsing here should be inserted in the document,
    // as scripts in event handlers, etc would be executed then.
    // we're only taking text, so that is fine
    const rootNode = new DOMParser().parseFromString(html, "text/html").body;
    const parts = [];
    let lastNode;
    let inQuote = isQuotedMessage;
    const state: IState = {
        listIndex: [],
    };

    function onNodeEnter(n) {
        if (checkIgnored(n)) {
            return false;
        }
        if (n.nodeName === "BLOCKQUOTE") {
            inQuote = true;
        }

        const newParts = [];
        if (lastNode && (checkBlockNode(lastNode) || checkBlockNode(n))) {
            newParts.push(partCreator.newline());
        }

        if (n.nodeType === Node.TEXT_NODE) {
            newParts.push(...parseAtRoomMentions(n.nodeValue, partCreator));
        } else if (n.nodeType === Node.ELEMENT_NODE) {
            const parseResult = parseElement(n, partCreator, lastNode, state);
            if (parseResult) {
                if (Array.isArray(parseResult)) {
                    newParts.push(...parseResult);
                } else {
                    newParts.push(parseResult);
                }
            }
        }

        if (newParts.length && inQuote) {
            const isFirstPart = parts.length === 0;
            prefixQuoteLines(isFirstPart, newParts, partCreator);
        }

        parts.push(...newParts);

        const descend = checkDescendInto(n);
        // when not descending (like for PRE), onNodeLeave won't be called to set lastNode
        // so do that here.
        lastNode = descend ? null : n;
        return descend;
    }

    function onNodeLeave(n) {
        if (checkIgnored(n)) {
            return;
        }
        switch (n.nodeName) {
            case "BLOCKQUOTE":
                inQuote = false;
                break;
            case "OL":
                state.listIndex.pop();
                /* falls through */
            case "UL":
                state.listDepth -= 1;
                break;
        }
        lastNode = n;
    }

    walkDOMDepthFirst(rootNode, onNodeEnter, onNodeLeave);

    return parts;
}

export function parsePlainTextMessage(body: string, partCreator: PartCreator, isQuotedMessage: boolean) {
    const lines = body.split(/\r\n|\r|\n/g); // split on any new-line combination not just \n, collapses \r\n
    return lines.reduce((parts, line, i) => {
        if (isQuotedMessage) {
            parts.push(partCreator.plain(QUOTE_LINE_PREFIX));
        }
        parts.push(...parseAtRoomMentions(line, partCreator));
        const isLast = i === lines.length - 1;
        if (!isLast) {
            parts.push(partCreator.newline());
        }
        return parts;
    }, []);
}

export function parseEvent(event: MatrixEvent, partCreator: PartCreator, {isQuotedMessage = false} = {}) {
    const content = event.getContent();
    let parts;
    if (content.format === "org.matrix.custom.html") {
        parts = parseHtmlMessage(content.formatted_body || "", partCreator, isQuotedMessage);
    } else {
        parts = parsePlainTextMessage(content.body || "", partCreator, isQuotedMessage);
    }
    if (content.msgtype === "m.emote") {
        parts.unshift(partCreator.plain("/me "));
    }
    return parts;
}
