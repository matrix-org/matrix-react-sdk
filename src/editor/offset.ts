/*
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import EditorModel from "./model";
import DocumentPosition from "./position";

export default class DocumentOffset {
    public constructor(public offset: number, public readonly atNodeEnd: boolean) {}

    public asPosition(model: EditorModel): DocumentPosition {
        return model.positionForOffset(this.offset, this.atNodeEnd);
    }

    public add(delta: number, atNodeEnd = false): DocumentOffset {
        return new DocumentOffset(this.offset + delta, atNodeEnd);
    }
}
