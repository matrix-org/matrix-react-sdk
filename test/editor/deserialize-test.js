/*
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import '../skinned-sdk'; // Must be first for skinning to work
import {parseEvent} from "../../src/editor/deserialize";
import {createPartCreator} from "./mock";

function htmlMessage(formattedBody, msgtype = "m.text") {
    return {
        getContent() {
            return {
                msgtype,
                format: "org.matrix.custom.html",
                formatted_body: formattedBody,
            };
        },
    };
}

function textMessage(body, msgtype = "m.text") {
    return {
        getContent() {
            return {
                msgtype,
                body,
            };
        },
    };
}

function mergeAdjacentParts(parts) {
    let prevPart;
    for (let i = 0; i < parts.length; ++i) {
        let part = parts[i];
        const isEmpty = !part.text.length;
        const isMerged = !isEmpty && prevPart && prevPart.merge(part);
        if (isEmpty || isMerged) {
            // remove empty or merged part
            part = prevPart;
            parts.splice(i, 1);
            //repeat this index, as it's removed now
            --i;
        }
        prevPart = part;
    }
}

function normalize(parts) {
    // merge adjacent parts as this will happen
    // in the model anyway, and whether 1 or multiple
    // plain parts are returned is an implementation detail
    mergeAdjacentParts(parts);
    // convert to data objects for easier asserting
    return parts.map(p => p.serialize());
}

describe('editor/deserialize', function() {
    describe('text messages', function() {
        it('test with newlines', function() {
            const parts = normalize(parseEvent(textMessage("hello\nworld"), createPartCreator()));
            expect(parts[0]).toStrictEqual({type: "plain", text: "hello"});
            expect(parts[1]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "world"});
            expect(parts.length).toBe(3);
        });
        it('@room pill', function() {
            const parts = normalize(parseEvent(textMessage("text message for @room"), createPartCreator()));
            expect(parts.length).toBe(2);
            expect(parts[0]).toStrictEqual({type: "plain", text: "text message for "});
            expect(parts[1]).toStrictEqual({type: "at-room-pill", text: "@room"});
        });
        it('emote', function() {
            const text = "says DON'T SHOUT!";
            const parts = normalize(parseEvent(textMessage(text, "m.emote"), createPartCreator()));
            expect(parts.length).toBe(1);
            expect(parts[0]).toStrictEqual({type: "plain", text: "/me says DON'T SHOUT!"});
        });
    });
    describe('html messages', function() {
        it('inline styling', function() {
            const html = "<strong>bold</strong> and <em>emphasized</em> text";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(1);
            expect(parts[0]).toStrictEqual({type: "plain", text: "**bold** and _emphasized_ text"});
        });
        it('hyperlink', function() {
            const html = 'click <a href="http://example.com/">this</a>!';
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(1);
            expect(parts[0]).toStrictEqual({type: "plain", text: "click [this](http://example.com/)!"});
        });
        it('multiple lines with paragraphs', function() {
            const html = '<p>hello</p><p>world</p>';
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(4);
            expect(parts[0]).toStrictEqual({type: "plain", text: "hello"});
            expect(parts[1]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[2]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[3]).toStrictEqual({type: "plain", text: "world"});
        });
        it('multiple lines with line breaks', function() {
            const html = 'hello<br>world';
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(3);
            expect(parts[0]).toStrictEqual({type: "plain", text: "hello"});
            expect(parts[1]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "world"});
        });
        it('multiple lines mixing paragraphs and line breaks', function() {
            const html = '<p>hello<br>warm</p><p>world</p>';
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(6);
            expect(parts[0]).toStrictEqual({type: "plain", text: "hello"});
            expect(parts[1]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "warm"});
            expect(parts[3]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[4]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[5]).toStrictEqual({type: "plain", text: "world"});
        });
        it('quote', function() {
            const html = '<blockquote><p><em>wise</em><br><strong>words</strong></p></blockquote><p>indeed</p>';
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(6);
            expect(parts[0]).toStrictEqual({type: "plain", text: "> _wise_"});
            expect(parts[1]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "> **words**"});
            expect(parts[3]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[4]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[5]).toStrictEqual({type: "plain", text: "indeed"});
        });
        it('user pill', function() {
            const html = "Hi <a href=\"https://matrix.to/#/@alice:hs.tld\">Alice</a>!";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(3);
            expect(parts[0]).toStrictEqual({type: "plain", text: "Hi "});
            expect(parts[1]).toStrictEqual({type: "user-pill", text: "Alice", resourceId: "@alice:hs.tld"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "!"});
        });
        it('user pill with displayname containing backslash', function() {
            const html = "Hi <a href=\"https://matrix.to/#/@alice:hs.tld\">Alice\\</a>!";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(3);
            expect(parts[0]).toStrictEqual({type: "plain", text: "Hi "});
            expect(parts[1]).toStrictEqual({type: "user-pill", text: "Alice\\", resourceId: "@alice:hs.tld"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "!"});
        });
        it('user pill with displayname containing opening square bracket', function() {
            const html = "Hi <a href=\"https://matrix.to/#/@alice:hs.tld\">Alice[[</a>!";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(3);
            expect(parts[0]).toStrictEqual({type: "plain", text: "Hi "});
            expect(parts[1]).toStrictEqual({type: "user-pill", text: "Alice[[", resourceId: "@alice:hs.tld"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "!"});
        });
        it('user pill with displayname containing closing square bracket', function() {
            const html = "Hi <a href=\"https://matrix.to/#/@alice:hs.tld\">Alice]</a>!";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(3);
            expect(parts[0]).toStrictEqual({type: "plain", text: "Hi "});
            expect(parts[1]).toStrictEqual({type: "user-pill", text: "Alice]", resourceId: "@alice:hs.tld"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "!"});
        });
        it('room pill', function() {
            const html = "Try <a href=\"https://matrix.to/#/#room:hs.tld\">#room:hs.tld</a>?";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(3);
            expect(parts[0]).toStrictEqual({type: "plain", text: "Try "});
            expect(parts[1]).toStrictEqual({type: "room-pill", text: "#room:hs.tld"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "?"});
        });
        it('@room pill', function() {
            const html = "<em>formatted</em> message for @room";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(2);
            expect(parts[0]).toStrictEqual({type: "plain", text: "_formatted_ message for "});
            expect(parts[1]).toStrictEqual({type: "at-room-pill", text: "@room"});
        });
        it('inline code', function() {
            const html = "there is no place like <code>127.0.0.1</code>!";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(1);
            expect(parts[0]).toStrictEqual({type: "plain", text: "there is no place like `127.0.0.1`!"});
        });
        it('code block with no trailing text', function() {
            const html = "<pre><code>0xDEADBEEF\n</code></pre>\n";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            console.log(parts);
            expect(parts.length).toBe(5);
            expect(parts[0]).toStrictEqual({type: "plain", text: "```"});
            expect(parts[1]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "0xDEADBEEF"});
            expect(parts[3]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[4]).toStrictEqual({type: "plain", text: "```"});
        });
        // failing likely because of https://github.com/vector-im/element-web/issues/10316
        xit('code block with no trailing text and no newlines', function() {
            const html = "<pre><code>0xDEADBEEF</code></pre>";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(5);
            expect(parts[0]).toStrictEqual({type: "plain", text: "```"});
            expect(parts[1]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "0xDEADBEEF"});
            expect(parts[3]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[4]).toStrictEqual({type: "plain", text: "```"});
        });
        it('unordered lists', function() {
            const html = "<ul><li>Oak</li><li>Spruce</li><li>Birch</li></ul>";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(5);
            expect(parts[0]).toStrictEqual({type: "plain", text: "- Oak"});
            expect(parts[1]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "- Spruce"});
            expect(parts[3]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[4]).toStrictEqual({type: "plain", text: "- Birch"});
        });
        it('ordered lists', function() {
            const html = "<ol><li>Start</li><li>Continue</li><li>Finish</li></ol>";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(5);
            expect(parts[0]).toStrictEqual({type: "plain", text: "1. Start"});
            expect(parts[1]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[2]).toStrictEqual({type: "plain", text: "2. Continue"});
            expect(parts[3]).toStrictEqual({type: "newline", text: "\n"});
            expect(parts[4]).toStrictEqual({type: "plain", text: "3. Finish"});
        });
        it('mx-reply is stripped', function() {
            const html = "<mx-reply>foo</mx-reply>bar";
            const parts = normalize(parseEvent(htmlMessage(html), createPartCreator()));
            expect(parts.length).toBe(1);
            expect(parts[0]).toStrictEqual({type: "plain", text: "bar"});
        });
        it('emote', function() {
            const html = "says <em>DON'T SHOUT</em>!";
            const parts = normalize(parseEvent(htmlMessage(html, "m.emote"), createPartCreator()));
            expect(parts.length).toBe(1);
            expect(parts[0]).toStrictEqual({type: "plain", text: "/me says _DON'T SHOUT_!"});
        });
    });
});
