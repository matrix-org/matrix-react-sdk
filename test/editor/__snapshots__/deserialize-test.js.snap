// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editor/deserialize html messages escapes angle brackets 1`] = `
Array [
  Object {
    "text": "\\\\> \\\\\\\\<del>no formatting here\\\\\\\\</del>",
    "type": "plain",
  },
]
`;

exports[`editor/deserialize html messages escapes asterisks 1`] = `
Array [
  Object {
    "text": "\\\\*hello\\\\*",
    "type": "plain",
  },
]
`;

exports[`editor/deserialize html messages escapes backslashes 1`] = `
Array [
  Object {
    "text": "C:\\\\\\\\My Documents",
    "type": "plain",
  },
]
`;

exports[`editor/deserialize html messages escapes backticks in code blocks 1`] = `
Array [
  Object {
    "text": "\`\`this â†’ \` is a backtick\`\`",
    "type": "plain",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "\`\`\`\`",
    "type": "plain",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "and here are 3 of them:",
    "type": "plain",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "\`\`\`",
    "type": "plain",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "\`\`\`\`",
    "type": "plain",
  },
]
`;

exports[`editor/deserialize html messages escapes backticks outside of code blocks 1`] = `
Array [
  Object {
    "text": "some \\\\\`backticks\\\\\`",
    "type": "plain",
  },
]
`;

exports[`editor/deserialize html messages escapes square brackets 1`] = `
Array [
  Object {
    "text": "\\\\[not an actual link\\\\](https://example.org)",
    "type": "plain",
  },
]
`;

exports[`editor/deserialize html messages escapes underscores 1`] = `
Array [
  Object {
    "text": "\\\\_\\\\_emphasis\\\\_\\\\_",
    "type": "plain",
  },
]
`;

exports[`editor/deserialize html messages preserves nested formatting 1`] = `
Array [
  Object {
    "text": "a<sub>b_c**d<u>e</u>**_</sub>",
    "type": "plain",
  },
]
`;

exports[`editor/deserialize html messages preserves nested quotes 1`] = `
Array [
  Object {
    "text": "> foo",
    "type": "plain",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "> ",
    "type": "plain",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "> > bar",
    "type": "plain",
  },
]
`;

exports[`editor/deserialize html messages surrounds lists with newlines 1`] = `
Array [
  Object {
    "text": "foo",
    "type": "plain",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "- bar",
    "type": "plain",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "
",
    "type": "newline",
  },
  Object {
    "text": "baz",
    "type": "plain",
  },
]
`;
